Try: Show drag slot instead of hiding the item being dragged.

jasmussen opened the pull request:
## Description

This is only an experiment.

At the moment, when you drag and drop blocks, the block you're dragging is "lifted" off the page, removed from its original spot, until dropped in its new spot. Like so:

![before](https://user-images.githubusercontent.com/1204802/128689702-304cc8b3-f096-4303-a3be-ebdb82e01fa9.gif)

This PR experimentally changes the behavior to keep a slot where you drag from:

![after](https://user-images.githubusercontent.com/1204802/128689800-885f6056-e19b-41bf-ad56-1f12f67ea21f.gif)

This is an experiment based on feedback that the current behavior is a little imprecise and prone to error. From this experiment, I feel primarily two take-aways:

1. The auto-scrolling of the viewport that is meant to make dragging long distances easier, is way too easy to trigger, making drag operations meant to just happen inside the visible part of the viewport far more likely to be janky.
2. There's a big opportunity to refactor the drag and drop method entirely, perhaps leverage Framer Motion to handle the shortcomings of the current system.

In lieu of a bigger refactor, this could be an interim fix provided it's believed to be better. I don't necessarily think it is, but since it's easy to try, I thought it worth it.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen committed:
Try: Show drag slot instead of hiding the item being dragged.

gwwar commented:
I think this definitely helps with the scroll content height changing. 🤔 If folks are interested in the idea, I'm guessing we might want more of a hint that it's a placeholder. The drop might be a little abrupt too without an animation, but that's probably fine.

apeatling commented:
This feels much better to me. The removal of the jumping of the content on the canvas helps the eye and brain not get lost where it's looking to drag. I think the transition animations and look of the drag slot could still improve, but this is a great start. 👍

apeatling commented:
Approving changes assuming you'd iterate in followups, but happy to re-review if you continue here.

jasmussen commented:
Thank you for the reviews, it's definitely affirming. Because it's a somewhat bigger change, visually, I'd love thoughts by @youknowriad or @mtias. There's also a chance this change will be moot or only passing if we tackle drag and drop in a more head-on/holistic way.

youknowriad commented on packages/block-editor/src/components/block-draggable/style.scss:
```diff
@@ -45,8 +45,18 @@
-// The effect is that the block being dragged appears to be "lifted".
+// This creates a "slot" where the block you're dragging appeared.
+// We use !important as one of the rules are meant to be overriden.
 .is-dragging {
```
That look like a very generic selector, it could affect a lot of things (including things not rendered by core Gutenberg), should we make it more specific?

jasmussen replied:
Good point. Let me see if I can scope it better, I know that would make @gwwar happy as well 😅

jasmussen replied:
Looks like I can scope it like so: `.block-editor-block-list__layout .is-dragging` — would that be preferable? It doesn't feel like the right file to store that particular rule for. 

The `!important`s still feel correct to me insofar as the way we make the "slot" appear is that we style and override the appearance of the block itself.

youknowriad replied:
For me `.block-editor-block-list__block.is-dragging` sounds like a better selector. Also, it seems `.is-dragging` class is used in the ListView as well, so we might want to check whether there's any regression there by these changes.

youknowriad reviewed (commented):
I personally like this new behavior. It avoid layout shifts.

mtias commented:
I don't have a strong opinion. If you are dragging a large section you need to scroll past it, which can be tedious.

jasmussen commented:
> If you are dragging a large section you need to scroll past it, which can be tedious.

We could potentially apply a max-height to the "slot". There'd still be a shift like there is today, but it'd only be in cases of tall blocks.

jasmussen committed:
Try max-height, and polish.

jasmussen commented:
Pushed a few tweaks, and tried to apply a max-height so tall items still collapse. However a small gotcha I discovered — when you multiselect. Since the is-dragging class that we are applying is applied on a per-block basis with no wrapping container when you're multiselecting, you get this:

![gotcha](https://user-images.githubusercontent.com/1204802/129188428-b8bf4eb4-4715-4ca2-9ed3-2f9299c3818f.gif)

I suppose that's expected, but something about that feels wrong to me. I could possibly write a complex selector that would make all those separate blocks appear as one, but it seems complex. What do you think?

apeatling reviewed (commented):
I wonder if after a certain height we could disable the drag slot, for example when it is more than 75% of the viewport height? This might alleviate situations where it's difficult to drag past the drag slot as Matías mentions.

mtias commented:
The max-height feels a bit awkward and arbitrary to me. It might make more sense to not render it all in those cases as mentioned.

jasmussen committed:
Remove max-height again.

jasmussen committed:
Hide subsequent multi selected blocks.

jasmussen commented:
Rebased and removed the max-height tweak from this one. I can't think of a simple way to make an "if taller than n pixels do this" behavior, so for the remaining issue around very tall blocks I explored these three options. 

1: do nothing. Just show multiple slots:

![drag-slot option 1](https://user-images.githubusercontent.com/1204802/130042982-afb772ab-ecd5-4e54-a687-76fb5c324693.gif)

I don't hate it. Combined with the viewport scrolling, it's technically a fairly accurate representation.

2: hide all but the first slot:

![drag-slot option 2](https://user-images.githubusercontent.com/1204802/130043078-00c906fc-c0c8-42ae-a42b-44efa601e81f.gif)

I like this one, and pushed it. It's not super accurate, but it feels like it still optimizes for the most common case.

3: hide multiselected blocks entirely:

![drag-slot option 3](https://user-images.githubusercontent.com/1204802/130043190-8e0f1503-2d54-4628-8a0b-6f7b75afef5e.gif)

Kind of like the current behavior, but only fur multiselected blocks.

Let me know your thoughts when you have a moment.

jasmussen commented:
Rebased this one to shepard it along. As a reminder, the open question is what behavior we should apply when dragging multiple blocks. This is what the branch here does currently:
![multi drag](https://user-images.githubusercontent.com/1204802/137087151-fda846ef-5e54-45c7-8a19-0a774f98c9f1.gif)

That is, every block after the first one is hidden, so that if you're dragging a ton of blocks, you don't leave a very tall "slot".

Coming back to this one after a while, I'm tempted to just accept that slot, and embrace this one:

<img src="https://user-images.githubusercontent.com/1204802/130042982-afb772ab-ecd5-4e54-a687-76fb5c324693.gif">

What do you think? And is this still something we should land?

apeatling reviewed (dismissed):
I think we should land this change by accepting that multiple blocks being dragged will show multiple slots and having no jump at all. If there's still a jump then it defeats a lot of the purpose of the change. Essentially what you have in the final gif above.

jasmussen commented:
Cool, I'll rebase and make the change tomorrow, then we can land it. That can also be considered notice for those subscribed to the ticket!

jasmussen committed:
Allow multiple sequential drag slots.

jasmussen commented:
I've changed the behavior back to work like the following, which shows a persisted drag slot for each item:

![dragondrop](https://user-images.githubusercontent.com/1204802/139387700-eab335e4-563b-4a94-a965-877ef425b192.gif)

I do like that it avoids layout shifts, that alone feels worth it to me to land this, and then we can continue looking at drag and drop improvements. For example the autoscroll behavior still feels too aggressive to me.

If you feel like apprpoving the PR, feel free to also merge it when you get the chance!

aristath reviewed (approved):
Feels a lot better than the current behavior.
Code is clean and makes perfect sense 👍

jasmussen commented:
Thanks, I'll land this, and I'm easy to find. There's so little code we can always rewind the clock if something happens!
